src/core/ipns/routing/pubsub-datastore.js

```diff
@@ -0,0 +1,167 @@
+'use strict'
+
+const ipns = require('ipns')
+const { fromB58String, toB58String } = require('multihashes')
+const PubsubDatastore = require('datastore-pubsub')
+
+const withIs = require('class-is')
+
+const errcode = require('err-code')
+const debug = require('debug')
+const log = debug('jsipfs:ipns:pubsub')
+log.error = debug('jsipfs:ipns:pubsub:error')
+
+// Pubsub datastore aims to manage the pubsub subscriptions for IPNS
+class IpnsPubsubDatastore {
+  constructor (pubsub, localDatastore, peerId) {
+    this._pubsub = pubsub
+    this._subscriptions = {}
+
+    // Bind _handleSubscriptionKey function, which is called by PubsubDatastore.
+    this._handleSubscriptionKey = this._handleSubscriptionKey.bind(this)
+    this._pubsubDs = new PubsubDatastore(pubsub, localDatastore, peerId, ipns.validator, this._handleSubscriptionKey)
+  }
+
+  /**
+   * Put a value to the pubsub datastore indexed by the received key properly encoded.
+   * @param {Buffer} key identifier of the value.
+   * @param {Buffer} value value to be stored.
+   * @param {function(Error)} callback
+   * @returns {void}
+   */
+  put (key, value, callback) {
+    if (!Buffer.isBuffer(key)) {
+      const errMsg = `key ${key} does not have a valid format`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_KEY'))
+    }
+
+    if (!Buffer.isBuffer(value)) {
+      const errMsg = `received value ${value} is not a buffer`
+
+      log.error(errMsg)
+      return callback(errcode(new Error(errMsg), 'ERR_INVALID_VALUE_RECEIVED'))
+    }
```

Comment:
You can remove these validations - `datastore-pubsub` does them and this method does not use `key` or `value` directly.